k8s/dialer.go

```diff
@@ -0,0 +1,324 @@
+package k8s
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+const (
+	socatImage = "quay.io/mvasek/socat:alpine"
+)
+
+type ContextDialer interface {
+	DialContext(ctx context.Context, network string, addr string) (net.Conn, error)
+	Close() error
+}
+
+// NewInClusterDialer creates context dialer that will dial TCP connections via POD running in k8s cluster.
+// This is useful when accessing k8s services that are not exposed outside cluster (e.g. openshift image registry).
+//
+// Usage:
+//
+//     dialer, err := k8s.NewInClusterDialer(ctx)
+//     if err != nil {
+//         return err
+//     }
+//     defer dialer.Close()
+//
+//     transport := &http.Transport{
+//         DialContext: dialer.DialContext,
+//     }
+//
+//     var client = http.Client{
+//         Transport: transport,
+//     }
+func NewInClusterDialer(ctx context.Context) (ContextDialer, error) {
+	c := &contextDialer{}
+	err := c.startDialerPod(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type contextDialer struct {
+	coreV1    v1.CoreV1Interface
+	restConf  *restclient.Config
+	podName   string
+	namespace string
+}
+
+func (c *contextDialer) DialContext(ctx context.Context, network string, addr string) (net.Conn, error) {
+	if !(network == "tcp" || network == "tcp4" || network == "tcp6") {
+		return nil, fmt.Errorf("unsupported network: %q", network)
+	}
+
+	execDone := make(chan struct{})
+	pr, pw, conn := newConn(execDone)
+
+	go func() {
+		defer close(execDone)
+		errOut := bytes.NewBuffer(nil)
+		err := c.exec(addr, pr, pw, errOut)
+		if err != nil {
+			err = fmt.Errorf("failed to exec in pod: %w (stderr: %q)", err, errOut.String())
+			_ = pr.CloseWithError(err)
+			_ = pw.CloseWithError(err)
+		}
+	}()
+
+	return conn, nil
+}
+
+func (c *contextDialer) Close() error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*1)
+	defer cancel()
+	delOpts := metaV1.DeleteOptions{}
+
+	return c.coreV1.Pods(c.namespace).Delete(ctx, c.podName, delOpts)
```

Comment:
You can do this if you are using a higher-level abstraction like `Deployment` or `ReplicaSet` if you want them to manage the lifecycle of your pod, otherwise, you have to do it on your own.

Reply:
btw, what do you mean with 'upon' completion ? When all containers in the pod stop ?

Reply:
Yes, when all processes exited.

Reply:
If the app for some reason crashes the `Close()` may not be called so there would be dangling completed pods.

Reply:
> @rhuss @markusthoemmes When I `exec` command in a pod it there a way to send a signal to the created process?

Maybe with a second `exec sh -c "cat /tmp/pid.txt | xargs kill"` (and writing the PID to a pid file in the first exec) ?